src/Model.php

```diff
@@ -0,0 +1,582 @@
+	 * @return bool
+	 */
+	 protected function performInsert(Query $query)
+	 {
+		 if (empty($this->attributes)) {
+			 return true;
+		 }
+
+		 $success = $query->insert();
+
+		 if ($success){
+			 $this->exists = true;
+		 }
+
+		 return $success;
+	 }
+
+	/**
+	 * Perform a model insert operation.
+	 *
+	 * @param  Query  $query
+	 * @return bool
+	 */
+	 protected function performUpdate($query)
+	 {
+		 if (empty($this->attributes)) {
+			 return true;
+		 }
+
+		 $success = $query->update();
+
+		 return $success;
+	 }
+
+	/**
+	 * Perform a model insert operation.
+	 *
+	 * @param  Query  $query
+	 * @return bool
+	 */
+	protected function performDelete($query)
+	{
+		$success = $query->delete($this);
+
+		if ($success)
+		{
+			$this->exists = false;
+		}
+
+		return $success;
+	}
+
+
+	/**
+	 * Get a new query builder for the model's table.
+	 *
+	 * @return Query
+	 */
+	public function newQuery()
+	{
+		return new Query($this->_dbFactory->getQuery($this->_db->getName(), $this->_db), $this->_db, $this);
+	}
+
+	/**
+	 * Handle dynamic method calls into the model.
+	 *
+	 * @param  string  $method
+	 * @param  array  $parameters
+	 * @return mixed
+	 */
+	public function __call($method, $parameters)
+	{
+		if (in_array($method, ['increment', 'decrement'])) {
+			return $this->$method(...$parameters);
+		}
+
+		return $this->newQuery()->$method(...$parameters);
+	}
+
+	/**
+	 * Handle dynamic static method calls into the method.
+	 *
+	 * @param  string  $method
+	 * @param  array  $parameters
+	 * @return mixed
+	 */
+	public static function __callStatic($method, $parameters)
+	{
+		return (new static)->$method(...$parameters);
+	}
+
+	/**
+	 * Create a new model instance that is existing.
+	 *
+	 * @param  array  $attributes
+	 * @param  string|null  $connection
+	 * @return static
+	 */
+	public function newFromBuilder($attributes = [], $connection = null)
+	{
+		$model = $this->newInstance([], true);
+
+		$model->setRawAttributes((array) $attributes, true);
+
+		$model->setDriver($connection ?: $this->getDriver());
+
+		return $model;
+	}
+
+	/**
+	 * Create a new instance of the given model.
+	 *
+	 * @param  array  $attributes
+	 * @param  bool  $exists
+	 * @return static
+	 */
+	public function newInstance($attributes = [], $exists = false)
+	{
+		// This method just provides a convenient way for us to generate fresh model
+		// instances of this current model. It is particularly useful during the
+		// hydration of new objects via the Eloquent query builder instances.
+		$model = new static((array) $attributes);
+
+		$model->exists = $exists;
+
+		$model->setDriver(
+			$this->getDriver()
+		);
+
+		return $model;
+	}
+
+
+	/**
+	 * sets the default value of the table name based on Model class name.
+	 */
+	private function setDefaultTable()
+	{
+		$className =  basename(str_replace('\\', '/', get_class($this)));
+
+		//TODO make a better way of constructing plurals.
+		$this->table = '#__' . strtolower($className) . 's';
```

Comment:
Import the string package and use the inflector class https://github.com/joomla-framework/string/blob/master/src/Inflector.php